src/TracerState.h

```diff
@@ -1339,25 +1339,39 @@ class TracerState {
     std::vector<unsigned int> object_count_;
     std::vector<std::pair<lifecycle_t, int>> lifecycle_summary_;
     std::vector<unsigned long int> event_counter_;
-
     std::vector<AssignmentState> assignment_stack_;
     
 public: 
-  void push_assignment_stack(SEXP symbol, SEXP env, sexptype_t type, SEXP calling_env) {
-    assignment_stack_.push_back(AssignmentState(symbol,env, type, calling_env));
-  }
-  
-  AssignmentState& peek_assignment_stack() {
-    return assignment_stack_[assignment_stack_.size() - 1];
-  }
-  
-  bool assignment_stack_is_empty() {
-    return assignment_stack_.size() == 0;
-  }
 
-void pop_assignment_stack() {
-    assignment_stack_.pop_back();
-  }
+    bool is_fresh_environment(SEXP rho) {
```

Comment:
This is probably not doing what you think it is doing. Just checking if the environment is not actively used by parent functions is not enough. What if the call finishes but its environment lives on?
We only want to discard cases where someone uses an environment whose parent is R_EmptyEnvironment.